File: solutions/kotlin/run-length-encoding/src/main/kotlin/RunLengthEncoding.kt
object RunLengthEncoding {

    fun encode(input: String): String {
        if (input == "") return ""
        val fst = input[0]
        val n = input.takeWhile { it == fst }
        val len = if (n.length > 1) n.length.toString() else ""
        return "${len}${fst}${encode(input.substring(n.length))}"
    }

    fun decode(input: String): String {
        if (input == "") return ""
        val digits = input.takeWhile { it.isDigit() }
        val n = if (digits.length == 0) 1 else Integer.parseInt(digits)
        val str = input.substring(digits.length, digits.length + 1).repeat(n)
        val rest = decode(input.substring(digits.length + 1))
        return "$str$rest"
    }
}

File: solutions/julia/run-length-encoding/run-length-encoding.jl
function encode(s)
    join(map(eachmatch(r"(.)\1*", s)) do m
        "$(length(m.match) > 1 ? length(m.match) : "")$(m.captures[1])"
    end)
end



function decode(s)
    join(map(eachmatch(r"(\d+)?(.)", s)) do m
        repeat(m.captures[2], parse(Int, @something m.captures[1] "1"))
    end)
end
